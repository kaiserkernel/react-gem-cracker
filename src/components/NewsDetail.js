import React, { Component } from 'react';
import { gql } from 'react-apollo';
import PropTypes from 'prop-types';
import url from 'url';

import convertNumberToTimeAgo from '../helpers/convertNumberToTimeAgo';

class NewsDetail extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    commentCount: PropTypes.number.isRequired,
    creationTime: PropTypes.number.isRequired,
    submitterId: PropTypes.string.isRequired,
    title: PropTypes.string,
    isPostScrutinyVisible: PropTypes.bool,
    points: PropTypes.number.isRequired,
    isFavoriteVisible: PropTypes.bool,
  }
  static defaultProps = {
    title: undefined,
    isFavoriteVisible: true,
    isPostScrutinyVisible: false,
  }

  upvote() {

  }
  hidestory() {

  }
  render() {
    return (
      <tr>
        <td colSpan="2" />
        <td className="subtext">
          <span className="score">{this.props.points} points</span>
          {' by '}
          <a href={`/user?id=${this.props.submitterId}`} className="hnuser">
            {this.props.submitterId}
          </a>
          {' '}
          <span className="age">
            <a href="item?id=15077449">
              {convertNumberToTimeAgo(this.props.creationTime)}
            </a>
          </span>
          {' | '}
          <a href="/hide?id=15077449&amp;goto=news&amp;auth=15140ad499d896ef90cc72930b3fb7706f6d6398" onClick={this.hidestory}>
          hide
          </a>
          {
            this.props.isPostScrutinyVisible &&
            <span>
              {' | '}
              <a href="https://hn.algolia.com/?query=Sublime%20Text%203.0&sort=byDate&dateRange=all&type=story&storyText=false&prefix&page=0" onClick={this.hidestory}>
              past
              </a>
              {' | '}
              <a href="https://www.google.com/search?q=Sublime%20Text%203.0" onClick={this.hidestory}>
              web
              </a>
            </span>
          }
          {' | '}
          <a href={`/item?id=${this.props.id}`}>
            {} {(() => {
              switch (this.props.commentCount) {
                case 0: return 'discuss';
                case 1: return '1 comment';
                default: return `${this.props.commentCount} comments`;
              }
            })()}
          </a>{this.props.isFavoriteVisible && ' | favorite'}
        </td>
      </tr>
    );
  }
}

const post = gql`
  fragment NewsDetail on entry {
    id
    submitterId
    commentCount
    points
    isFavoriteVisible
  }
`

export default NewsDetail;
