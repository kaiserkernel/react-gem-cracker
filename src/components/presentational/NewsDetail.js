import React, { Component } from 'react';
import { gql, graphql } from 'react-apollo';
import Link from 'next/link';
import PropTypes from 'prop-types';
import url from 'url';


import convertNumberToTimeAgo from '../../helpers/convertNumberToTimeAgo';

export default class NewsDetail extends Component {
  static propTypes = {
    id: PropTypes.number.isRequired,
    commentCount: PropTypes.number.isRequired,
    creationTime: PropTypes.number.isRequired,
    hidden: PropTypes.bool.isRequired,
    hideNewsItem: PropTypes.func.isRequired,
    isPostScrutinyVisible: PropTypes.bool,
    isFavoriteVisible: PropTypes.bool,
    isJobListing: PropTypes.bool,
    submitterId: PropTypes.string.isRequired,
    upvoteCount: PropTypes.number.isRequired,
  }
  static defaultProps = {
    isFavoriteVisible: true,
    isPostScrutinyVisible: false,
    isJobListing: false,
  }
  static fragments = {
    newsItem: gql`
      fragment NewsDetail on NewsItem {
        id
        commentCount
        creationTime
        hidden
        submitterId
        upvoteCount
      }
    `,
  };
  render() {
    return (
      this.props.isJobListing ?
        <tr>
          <td colSpan="2" />
          <td className="subtext">
            <span className="age">
              <Link prefetch href={`/item?id=${this.props.id}`}>
                <a>
                  {convertNumberToTimeAgo(this.props.creationTime)}
                </a>
              </Link>
            </span>
          </td>
        </tr>
        :
        <tr>
          <td colSpan="2" />
          <td className="subtext">
            <span className="score">{this.props.upvoteCount} points</span>
            {' by '}
            <Link prefetch href={`/user?id=${this.props.submitterId}`}>
              <a className="hnuser">
                {this.props.submitterId}
              </a>
            </Link>
            {' '}
            <span className="age">
              <Link prefetch href={`/item?id=${this.props.id}`}>
                <a>
                  {convertNumberToTimeAgo(this.props.creationTime)}
                </a>
              </Link>
            </span>
            {' | '}
            {
              this.props.hidden ?
                <a href="javascript:void(0)" onClick={() => this.props.hideNewsItem(this.props.id)}>
                  hide
                </a>
                :
                <a href="javascript:void(0)" onClick={() => this.props.unhideNewsItem(this.props.id)}>
                  hide
                </a>
            }
            {
              this.props.isPostScrutinyVisible &&
              <span>
                {' | '}
                <a href="https://hn.algolia.com/?query=Sublime%20Text%203.0&sort=byDate&dateRange=all&type=story&storyText=false&prefix&page=0">
                past
                </a>
                {' | '}
                <a href="https://www.google.com/search?q=Sublime%20Text%203.0">
                web
                </a>
              </span>
            }
            {' | '}
            <Link prefetch href={`/item?id=${this.props.id}`}>
              <a>
                {(() => {
                  switch (this.props.commentCount) {
                    case 0: return 'discuss';
                    case 1: return '1 comment';
                    default: return `${this.props.commentCount} comments`;
                  }
                })()}
              </a>
            </Link>
            {this.props.isFavoriteVisible && ' | favorite'}
          </td>
        </tr>
    );
  }
}
