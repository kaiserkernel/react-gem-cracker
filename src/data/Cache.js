// Interface: Commentable (Object can be commented on) comments, commentCount, commenter
// Interface: Voteable (Object can be voted on) upvotes, upvoteCount, hidden, hiddenCount,
// Data Type: Comment, can be on a news item or another comment
// Every time an upvote/downvote/comment is made, update the count

import LRU from 'lru-cache';

import {
  Feed,
} from './models';

// The cache is a singleton

export function warmCache() {
  // Fetch the front pages
  Feed.getForType('HOT', 30, 0);
  Feed.getForType('NEW', 30, 0);
  Feed.getForType('TOP', 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */


  getNewsItem(id) {
    // return this.newsItems.find(newsItem => newsItem.id === id);
    return this.newsItemsCache.get(id);
  }
  setNewsItem(newsItem) {
    return this.newsItemsCache.set(newsItem.id, newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */


  getUser(id) {
    return this.userCache.get(id);
  }

  getUsers() {
    return this.userCache.dump();
  }

  createUser(user) {
    this.userCache.set(user.id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN CACHES                         */


  newNewsItemsCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  newsItemsCache = LRU({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  })
  userCache = LRU({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  })

  /*                   END CACHES                         */
}

export default new Cache();
