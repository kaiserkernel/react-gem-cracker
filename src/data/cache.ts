import { debug } from 'debug';
import LRU from 'lru-cache';

import { FeedSingleton, NewsItem } from './models';
import { FeedType } from './models/feed';
import { User } from './models/user';
import { Comment } from './models/comment';

const logger = debug('app:Cache');
logger.log = console.log.bind(console);

// The cache is a singleton

export function warmCache(): void {
  // Fetch the front pages
  FeedSingleton.getForType(FeedType.TOP, 30, 0);
  FeedSingleton.getForType(FeedType.NEW, 30, 0);

  // Run every 15 mins
  setTimeout(warmCache, 1000 * 60 * 15);
}

class Cache {
  isReady = false;

  /*                  BEGIN NEWS ITEMS                      */

  getNewsItem(id: number): NewsItem | undefined {
    return this.newsItemsCache.get(id.toString());
  }

  setNewsItem(id: number, newsItem: NewsItem): boolean {
    return this.newsItemsCache.set(id.toString(), newsItem);
  }

  /*                  END NEWS ITEMS                      */

  /*                   BEGIN USERS                        */

  getUser(id): User | undefined {
    return this.userCache.get(id);
  }

  getUsers(): Array<LRU.Entry<string, User>> {
    return this.userCache.dump();
  }

  setUser(id: string, user: User): User {
    logger(`Cache set user ${user}`);
    this.userCache.set(id, user);
    return user;
  }

  /*                    END USERS                         */

  /*                   BEGIN COMMENTS                        */

  getComment(id: number): Comment | undefined {
    return this.commentCache.get(id.toString());
  }

  setComment(id: number, comment: Comment): Comment {
    this.commentCache.set(comment.id.toString(), comment);
    logger(`Cache set comment ${comment}`);
    return comment;
  }

  /*                    END COMMENTS                         */

  /*                   BEGIN CACHES                         */

  newNewsItemsCache = new LRU<string, NewsItem>({
    max: 500,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  newsItemsCache = new LRU<string, NewsItem>({
    max: 1000,
    maxAge: 1000 * 60 * 60, // 60 Minute cache: ms * s * m
  });

  userCache = new LRU<string, User>({
    max: 500,
    maxAge: 1000 * 60 * 60 * 2, // 2 hour cache: ms * s * m
  });

  commentCache = new LRU<string, Comment>({
    max: 5000,
    maxAge: 1000 * 60 * 60 * 1, // 1 hour cache: ms * s * m
  });

  /*                   END CACHES                         */
}

export const cache = new Cache();
