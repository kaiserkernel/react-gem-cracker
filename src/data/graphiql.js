/**  
   * WARNING: THIS USES THE REAL SCHEMA AND DATABASE!
   *
   *   Take care not to cause data loss but the app should be secure
   */

// TODO: Can make this graphql server part of the main server using feature flag
// Example GraphiQL server https://github.com/graphql/graphiql/tree/master/example

import express from 'express';
import bodyParser from 'body-parser';
import {
  graphqlExpress,
  graphiqlExpress,
} from 'apollo-server-express';

import Schema from './Schema';
import {
  APP_PORT,
  graphQLPath,
  graphiQLPath,
  GRAPHQL_URL,
  GRAPHIQL_URL,
} from '../config';
import {
  Comment,
  Feed,
  NewsItem,
  User,
} from './models';


const logger = console; // global.logger;

const app = express();
// app.use(morgan('dev'));

// Mount GraphQL Server middleware
app.use(graphQLPath, bodyParser.json(), graphqlExpress(request => ({
  schema: Schema,
  rootValue: { request },
  context: {
    Comment,
    Feed,
    NewsItem,
    User,
  },
  debug: true,
})));

app.use(graphiQLPath, graphiqlExpress({
  endpointURL: graphQLPath,
}));

app.listen(APP_PORT, () => {
  logger.info(`==> ðŸ“ˆ GraphQL Server on ${GRAPHQL_URL}`);
  logger.info(`==> ðŸŒŽ  Go to ${GRAPHIQL_URL}`);
});
