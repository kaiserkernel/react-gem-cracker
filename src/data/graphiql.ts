// /**
//  * WARNING: THIS USES THE REAL SCHEMA AND DATABASE!
//  *
//  *   Take care not to cause data loss but the app should be secure
//  */

// // TODO: Can make this graphql server part of the main server using feature flag
// // Example GraphiQL server https://github.com/graphql/graphiql/tree/master/example

// import { graphiqlExpress, graphqlExpress } from 'apollo-server-express';
// import * as bodyParser from 'body-parser';
// import * as express from 'express';
// import { Comment, FeedSingleton, NewsItem, User } from './models';
// import { schema } from './schema';
// import { APP_PORT, GRAPHIQL_URL, graphiQLPath, GRAPHQL_URL, graphQLPath } from '../config';

// const logger = console;

// const app = express();

// // Mount GraphQL Server middleware
// app.use(
//   graphQLPath,
//   bodyParser.json(),
//   graphqlExpress(request => ({
//     schema,
//     rootValue: { request },
//     context: {
//       Comment,
//       Feed: FeedSingleton,
//       NewsItem,
//       User,
//     },
//     debug: true,
//   }))
// );

// app.use(
//   graphiQLPath,
//   graphiqlExpress({
//     endpointURL: graphQLPath,
//   })
// );

// app.listen(APP_PORT, () => {
//   logger.info(`==> ðŸ“ˆ GraphQL Server on ${GRAPHQL_URL}`);
//   logger.info(`==> ðŸŒŽ  Go to ${GRAPHIQL_URL}`);
// });
