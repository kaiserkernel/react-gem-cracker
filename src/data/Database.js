import data from './SampleData';
import cache from './Cache';

/*                  BEGIN NEWS ITEMS                      */


export function getNewsItem(id) {
  return data.newsItems.find(newsItem => newsItem.id === id);
}

export function rankNewsItems() {
  // Would implement this somewhere in the real HN system
  // A scheduled job to recalculate post rank and points
  // .sort((a, b) => (a.rank - b.rank))
}

export function createNewsItem(newsItem) {
  data.newsItems.push(newsItem);
  return newsItem;
}

//                  NEWS ITEM MUTATIONS


export function upvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.upvotes.add(userId);
  newsItemData.upvoteCount += 1;

  return newsItemData;
}

export function downvoteNewsItem(id, userId) {
  const newsItemData = data.newsItems.find(newsItem => newsItem.id === id);
  newsItemData.downvotes.add(userId);
  newsItemData.downvoteCount += 1;

  return newsItemData;
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */


export function getNewNewsItems(first, skip) {
  return data.new.slice(skip, skip + first)
    .map((postId, index) => ({
      ...getNewsItem(postId),
      rank: skip + index + 1,
    }));
}

export function getTopNewsItems(first, skip) {
  return data.newsItems.slice(skip, skip + first);
}

export function getHotNews() {
  return data.newsItems;
}

export function getNewsItems() {
  return data.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */


export function getUser(id) {
  return data.users.find(user => user.id === id);
}

export function getUsers() {
  return data.users;
}

export function createUser(user) {
  data.users.push(user);
  return user;
}

/*                    END USERS                         */
