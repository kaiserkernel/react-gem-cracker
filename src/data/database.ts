import { debug } from 'debug';

import { cache } from './cache';
import { FeedSingleton, User } from './models';
import { sampleData } from './sample-data';
import { NewsItem } from './models/news-item';

const logger = debug('app:Database');
logger.log = console.log.bind(console);

/*                  BEGIN NEWS ITEMS                      */

export function getNewsItem(id: number): NewsItem | undefined {
  return sampleData.newsItems.find((newsItem) => newsItem.id === id);
}

export function createNewsItem(newsItem: NewsItem): NewsItem {
  sampleData.newsItems.push(newsItem);

  return newsItem;
}

//                  NEWS ITEM MUTATIONS

export function upvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  // Upvote the News Item in the DB
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.push(userId);
    newsItem.upvoteCount += 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function unvoteNewsItem(id: number, userId: string): NewsItem | undefined {
  const newsItem = cache.getNewsItem(id);

  if (newsItem && !newsItem.upvotes.includes(userId)) {
    newsItem.upvotes.splice(newsItem.upvotes.indexOf(userId), 1);
    newsItem.upvoteCount -= 1;
    cache.setNewsItem(id, newsItem);
  }

  return newsItem;
}

export function hideNewsItem(id: number, userId: string): NewsItem {
  logger(`Hiding News Item ${id} by ${userId}`);

  const newsItem = cache.getNewsItem(id);
  const user = cache.getUser(userId);

  if (user && !user.hides.includes(id) && newsItem && !newsItem.hides.includes(userId)) {
    user.hides.push(id);
    cache.setUser(userId, user);

    newsItem.hides.push(userId);
    cache.setNewsItem(id, newsItem);

    logger(`Hid News Item ${id} by ${userId}`);
  } else {
    throw new Error(`Data error, user has already hidden ${id} by ${userId}`);
  }

  return newsItem;
}

export function submitNewsItem(id: number, newsItem: NewsItem) {
  // Submit the News Item in the DB
  if (cache.setNewsItem(id, newsItem)) {
    FeedSingleton.new.unshift(id);
    FeedSingleton.new.pop();
    return newsItem;
  }

  throw new Error('Unable to submit News Item.');
}

/*                  END NEWS ITEMS                      */

/*                     BEGIN FEED                         */

export function getNewNewsItems(first: number, skip: number) {
  return sampleData.new.slice(skip, skip + first).map((postId, index) => ({
    ...getNewsItem(postId),
    rank: skip + index + 1,
  }));
}

export function getTopNewsItems(first: number, skip: number) {
  return sampleData.newsItems.slice(skip, skip + first);
}

export function getHotNews(): NewsItem[] {
  return sampleData.newsItems;
}

export function getNewsItems(): NewsItem[] {
  return sampleData.newsItems;
}

/*                     END FEED                         */

/*                   BEGIN USERS                        */

export function getUser(id: string) {
  return sampleData.users.find((user) => user.id === id);
}

export function getUsers() {
  return sampleData.users;
}

export function createUser(user: User) {
  sampleData.users.push(user);

  return user;
}

/*                    END USERS                         */
