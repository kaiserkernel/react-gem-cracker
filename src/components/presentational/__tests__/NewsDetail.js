import React from 'react';
import MockDate from 'mockdate';
// import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import NewsDetail from '../NewsDetail';
import data from '../../../data/SampleData';

const newsItem = data.newsItems[0];
// Snapshot will be out of date if we don't use consistent time agoy
// newsItem.creationTime = new Date().valueOf();
MockDate.set(1506022129802);

describe('NewsFeed component', () => {
  it('renders news items passed in as props', () => {
    const hideNewsItem = () => console.log('1');
    const wrapper = shallow(
      <NewsDetail {...newsItem} hideNewsItem={hideNewsItem} isFavoriteVisible={false} />,
    );
    expect(wrapper).toMatchSnapshot();
  });
});
