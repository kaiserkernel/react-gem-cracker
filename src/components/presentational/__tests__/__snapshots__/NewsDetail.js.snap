// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsFeed component renders news items passed in as props 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <tr>
    <td
        colSpan="2"
    />
    <td
        className="subtext"
    >
        <span
            className="score"
        >
            171
             points
        </span>
         by 
        <Link
            href="/user?id=hvo"
            prefetch={true}
        >
            <a
                className="hnuser"
            >
                hvo
            </a>
        </Link>
         
        <span
            className="age"
        >
            <Link
                href="/item?id=1224"
                prefetch={true}
            >
                <a>
                    14 days ago
                </a>
            </Link>
        </span>
         | 
        <a
            href="javascript:void(0)"
            onClick={[Function]}
        >
            hide
        </a>
         | 
        <Link
            href="/item?id=1224"
            prefetch={true}
        >
            <a>
                44 comments
            </a>
        </Link>
    </td>
</tr>,
  "nodes": Array [
    <tr>
      <td
            colSpan="2"
      />
      <td
            className="subtext"
      >
            <span
                  className="score"
            >
                  171
                   points
            </span>
             by 
            <Link
                  href="/user?id=hvo"
                  prefetch={true}
            >
                  <a
                        className="hnuser"
                  >
                        hvo
                  </a>
            </Link>
             
            <span
                  className="age"
            >
                  <Link
                        href="/item?id=1224"
                        prefetch={true}
                  >
                        <a>
                              14 days ago
                        </a>
                  </Link>
            </span>
             | 
            <a
                  href="javascript:void(0)"
                  onClick={[Function]}
            >
                  hide
            </a>
             | 
            <Link
                  href="/item?id=1224"
                  prefetch={true}
            >
                  <a>
                        44 comments
                  </a>
            </Link>
      </td>
</tr>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <NewsDetail
        commentCount={44}
        comments={Array []}
        creationTime={1504740146516}
        hidden={false}
        hiddenCount={0}
        hideNewsItem={[Function]}
        hides={Array []}
        id={1224}
        isFavoriteVisible={false}
        isJobListing={false}
        isPostScrutinyVisible={false}
        submitterId="hvo"
        text={null}
        title="“Learning How to Learn,” the most popular course on Coursera"
        upvoteCount={171}
        upvoted={false}
        upvotes={Array []}
        url="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NewsDetail {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "commentCount": 44,
          "comments": Array [],
          "creationTime": 1504740146516,
          "hidden": false,
          "hiddenCount": 0,
          "hideNewsItem": [Function],
          "hides": Array [],
          "id": 1224,
          "isFavoriteVisible": false,
          "isJobListing": false,
          "isPostScrutinyVisible": false,
          "submitterId": "hvo",
          "text": null,
          "title": "“Learning How to Learn,” the most popular course on Coursera",
          "upvoteCount": 171,
          "upvoted": false,
          "upvotes": Array [],
          "url": "https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well",
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <tr>
          <td
                    colSpan="2"
          />
          <td
                    className="subtext"
          >
                    <span
                              className="score"
                    >
                              171
                               points
                    </span>
                     by 
                    <Link
                              href="/user?id=hvo"
                              prefetch={true}
                    >
                              <a
                                        className="hnuser"
                              >
                                        hvo
                              </a>
                    </Link>
                     
                    <span
                              className="age"
                    >
                              <Link
                                        href="/item?id=1224"
                                        prefetch={true}
                              >
                                        <a>
                                                  14 days ago
                                        </a>
                              </Link>
                    </span>
                     | 
                    <a
                              href="javascript:void(0)"
                              onClick={[Function]}
                    >
                              hide
                    </a>
                     | 
                    <Link
                              href="/item?id=1224"
                              prefetch={true}
                    >
                              <a>
                                        44 comments
                              </a>
                    </Link>
          </td>
</tr>,
        "_debugID": 2,
        "_renderedOutput": <tr>
          <td
                    colSpan="2"
          />
          <td
                    className="subtext"
          >
                    <span
                              className="score"
                    >
                              171
                               points
                    </span>
                     by 
                    <Link
                              href="/user?id=hvo"
                              prefetch={true}
                    >
                              <a
                                        className="hnuser"
                              >
                                        hvo
                              </a>
                    </Link>
                     
                    <span
                              className="age"
                    >
                              <Link
                                        href="/item?id=1224"
                                        prefetch={true}
                              >
                                        <a>
                                                  14 days ago
                                        </a>
                              </Link>
                    </span>
                     | 
                    <a
                              href="javascript:void(0)"
                              onClick={[Function]}
                    >
                              hide
                    </a>
                     | 
                    <Link
                              href="/item?id=1224"
                              prefetch={true}
                    >
                              <a>
                                        44 comments
                              </a>
                    </Link>
          </td>
</tr>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <NewsDetail
    commentCount={44}
    comments={Array []}
    creationTime={1504740146516}
    hidden={false}
    hiddenCount={0}
    hideNewsItem={[Function]}
    hides={Array []}
    id={1224}
    isFavoriteVisible={false}
    isJobListing={false}
    isPostScrutinyVisible={false}
    submitterId="hvo"
    text={null}
    title="“Learning How to Learn,” the most popular course on Coursera"
    upvoteCount={171}
    upvoted={false}
    upvotes={Array []}
    url="https://www.nytimes.com/2017/08/04/education/edlife/learning-how-to-learn-barbara-oakley.html?hpw&rref=education&action=click&pgtype=Homepage&module=well-region&region=bottom-well&WT.nav=bottom-well"
/>,
}
`;
